src/Stack/Build/ConstructPlan.hs

```diff
@@ -220,20 +217,51 @@ constructPlan ls0 baseConfigOpts0 locals extraToBuild0 localDumpPkgs loadPackage
             prettyErrorNoIndent $ pprintExceptions errs stackYaml stackRoot parents (wanted ctx)
             throwM $ ConstructPlanFailed "Plan construction failed."
   where
-    hasBaseInDeps bconfig = Map.member (mkPackageName "base") (bcDependencies bconfig)
+    hasBaseInDeps = Map.member (mkPackageName "base") (smDeps sourceMap)
 
-    mkCtx econfig = Ctx
-        { ls = ls0
-        , baseConfigOpts = baseConfigOpts0
+    mkCtx econfig sources = Ctx
+        { baseConfigOpts = baseConfigOpts0
         , loadPackage = \x y z -> runRIO econfig $ loadPackage0 x y z
-        , combinedMap = combineMap sourceMap installedMap
+        , combinedMap = combineMap sources installedMap
         , ctxEnvConfig = econfig
         , callStack = []
-        , extraToBuild = extraToBuild0
-        , wanted = wantedLocalPackages locals <> extraToBuild0
-        , localNames = Set.fromList $ map (packageName . lpPackage) locals
+        , wanted = Map.keysSet (smtTargets $ smTargets sourceMap)
+        , localNames = Map.keysSet (smProject sourceMap) -- Set.fromList $ map (packageName . lpPackage) locals
         }
 
+    getSources = do
+      pPackages <- for (smProject sourceMap) $ \pp -> do
+        lp <- loadLocalPackage sourceMap pp
+        return $ PSFilePath lp Local
+      bopts <- view $ configL.to configBuild
+      env <- ask
+      let buildHaddocks = shouldHaddockDeps bopts
+          globalDeps = Map.mapMaybeWithKey globalToSource $ smGlobal sourceMap
+          globalToSource name gp | name `Set.member` wiredInPackages = Nothing
+                                 | otherwise =
+            let version = gpVersion gp
+                loc = PLIHackage (PackageIdentifierRevision name version CFILatest) Nothing
+                common = CommonPackage
+                  { cpGPD = runRIO env $ loadCabalFile (PLImmutable loc)
+                  , cpName = name
+                  , cpFlags = mempty
+                  , cpGhcOptions = mempty
+                  , cpHaddocks = buildHaddocks
+                  }
+            in Just $ PSRemote loc version NotFromSnapshot common
+      deps <- for (smDeps sourceMap) $ \dp ->
+        case dpLocation dp of
+          PLImmutable loc -> do
+            version <- getPLIVersion loc (cpGPD $ dpCommon dp)
+            return $ PSRemote loc version (dpFromSnapshot dp) (dpCommon dp)
+          PLMutable dir -> do
+            -- FIXME this is not correct, we don't want to treat all Mutable as local
```

Comment:
@qrilka this is the FIXME I had in mind when talking about an immutable package becoming mutable if it depends on some local mutable package, but here when it would become mutable it wouldn't be local and wouldn't have a local path.

Reply:
Cool, thanks